File: src/config.test.ts
import { describe, expect, test } from 'bun:test';
import { homedir } from 'node:os';
import { join } from 'node:path';
import { DEFAULT_TODO_DIR, DEFAULT_TODO_PORT, resolveTodoRuntimeConfig } from './config';
import { DEFAULT_STATUSLINE_TEMPLATE } from './statusline';

describe('resolveTodoRuntimeConfig', () => {
  test('defaults when nothing is set', () => {
    const resolved = resolveTodoRuntimeConfig({});
    expect(resolved.port).toBe(DEFAULT_TODO_PORT);
    expect(resolved.dir).toBe(DEFAULT_TODO_DIR);
  });

  test('rocky.json todo block overrides defaults', () => {
    const resolved = resolveTodoRuntimeConfig({}, { port: 9000, dir: '/data/todo' });
    expect(resolved.port).toBe(9000);
    expect(resolved.dir).toBe('/data/todo');
  });

  test('env wins over config, tilde is expanded', () => {
    const resolved = resolveTodoRuntimeConfig(
      { ROCKY_TODO_PORT: '9999', ROCKY_TODO_DIR: '~/custom' },
      { port: 9000, dir: '/data/todo' },
    );
    expect(resolved.port).toBe(9999);
    expect(resolved.dir).toBe(join(homedir(), 'custom'));
  });

  test('invalid env port falls through to config/default', () => {
    expect(resolveTodoRuntimeConfig({ ROCKY_TODO_PORT: 'abc' }).port).toBe(DEFAULT_TODO_PORT);
    expect(resolveTodoRuntimeConfig({ ROCKY_TODO_PORT: '-1' }, { port: 9000 }).port).toBe(9000);
  });

  test('expose defaults to empty (loopback bind, no channels)', () => {
    const resolved = resolveTodoRuntimeConfig({});
    expect(resolved.expose).toEqual([]);
    expect(resolved.host).toBe('127.0.0.1');
  });

  test('lan channel binds all interfaces; tailscale alone keeps loopback; both combine', () => {
    const lan = resolveTodoRuntimeConfig({}, { expose: ['lan'] });
    expect(lan.host).toBe('0.0.0.0');

    const ts = resolveTodoRuntimeConfig({}, { expose: ['tailscale-serve'] });
    expect(ts.expose).toEqual(['tailscale-serve']);
    expect(ts.host).toBe('127.0.0.1');

    const both = resolveTodoRuntimeConfig({}, { expose: ['lan', 'tailscale-serve'] });
    expect(both.expose).toEqual(['lan', 'tailscale-serve']);
    expect(both.host).toBe('0.0.0.0');

    // 문자열 하나도 배열로 정규화
    const single = resolveTodoRuntimeConfig({}, { expose: 'lan' });
    expect(single.expose).toEqual(['lan']);
    expect(single.host).toBe('0.0.0.0');

    // "off" / null 은 미설정과 동일
    expect(resolveTodoRuntimeConfig({}, { expose: 'off' }).expose).toEqual([]);
    expect(resolveTodoRuntimeConfig({}, { expose: null }).expose).toEqual([]);
  });

  test('env ROCKY_TODO_EXPOSE is comma-separated and wins entirely when set', () => {
    expect(resolveTodoRuntimeConfig({ ROCKY_TODO_EXPOSE: 'lan,tailscale-serve' }).expose).toEqual([
      'lan',
      'tailscale-serve',
    ]);
    // env 가 설정돼 있으면 config 보다 무조건 우선 — "off" 로 강제 차단 가능
    expect(
      resolveTodoRuntimeConfig({ ROCKY_TODO_EXPOSE: 'off' }, { expose: ['lan'] }).expose,
    ).toEqual([]);
    expect(
      resolveTodoRuntimeConfig({ ROCKY_TODO_EXPOSE: 'banana' }, { expose: ['lan'] }).expose,
    ).toEqual([]);
  });
});

describe('resolveTodoRuntimeConfig — statuslineTemplate', () => {
  test('설정이 없으면 기본 템플릿', () => {
    expect(resolveTodoRuntimeConfig({}, {}).statuslineTemplate).toBe(DEFAULT_STATUSLINE_TEMPLATE);
  });

  test('config > 기본', () => {
    const config = { statusline: { template: '[{doing}]' } };
    expect(resolveTodoRuntimeConfig({}, config).statuslineTemplate).toBe('[{doing}]');
  });

  test('env > config', () => {
    const config = { statusline: { template: '[{doing}]' } };
    const env = { ROCKY_TODO_STATUSLINE: '[{stale}]' };
    expect(resolveTodoRuntimeConfig(env, config).statuslineTemplate).toBe('[{stale}]');
  });

  test('빈 문자열은 오설정으로 보고 기본값으로 떨어진다', () => {
    // 끄고 싶으면 statusline 명령에서 curl 을 빼면 된다 — 여기서 빈 값을 통과시키면
    // 왜 아무것도 안 뜨는지 알아낼 방법이 없다.
    const env = { ROCKY_TODO_STATUSLINE: '   ' };
    expect(resolveTodoRuntimeConfig(env, { statusline: { template: '' } }).statuslineTemplate).toBe(
      DEFAULT_STATUSLINE_TEMPLATE,
    );
  });
});

File: crates/rocky-todo-core/tests/config_test.rs
//! TS `src/config.test.ts` + `src/rocky-config.test.ts` 포팅.

use rocky_todo_core::config::*;
use rocky_todo_core::statusline::DEFAULT_STATUSLINE_TEMPLATE;

fn env(pairs: &[(&str, &str)]) -> EnvMap {
    pairs
        .iter()
        .map(|(k, v)| (k.to_string(), v.to_string()))
        .collect()
}

#[test]
fn defaults_when_nothing_is_set() {
    let runtime = resolve_runtime_config(&env(&[]), &TodoConfig::default());
    assert_eq!(runtime.port, 8636);
    assert_eq!(runtime.host, "127.0.0.1");
    assert!(runtime.expose.is_empty());
    assert!(runtime
        .dir
        .to_string_lossy()
        .ends_with(".config/rocky/todo"));
}

#[test]
fn todo_block_overrides_defaults() {
    let todo = TodoConfig {
        port: Some(9000),
        dir: Some("/data/todo".into()),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[]), &todo);
    assert_eq!(runtime.port, 9000);
    assert_eq!(runtime.dir.to_string_lossy(), "/data/todo");
}

#[test]
fn env_wins_and_tilde_expands() {
    let todo = TodoConfig {
        port: Some(9000),
        dir: Some("/data/todo".into()),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(
        &env(&[
            ("ROCKY_TODO_PORT", "9100"),
            ("ROCKY_TODO_DIR", "~/todo-dir"),
        ]),
        &todo,
    );
    assert_eq!(runtime.port, 9100);
    assert!(runtime.dir.is_absolute());
    assert!(runtime.dir.to_string_lossy().ends_with("todo-dir"));
    assert!(!runtime.dir.to_string_lossy().contains('~'));
}

#[test]
fn invalid_env_port_falls_through() {
    let todo = TodoConfig {
        port: Some(9000),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[("ROCKY_TODO_PORT", "abc")]), &todo);
    assert_eq!(runtime.port, 9000);
    let runtime = resolve_runtime_config(&env(&[("ROCKY_TODO_PORT", "0")]), &TodoConfig::default());
    assert_eq!(runtime.port, 8636);
}

#[test]
fn expose_defaults_to_loopback() {
    let runtime = resolve_runtime_config(&env(&[]), &TodoConfig::default());
    assert!(runtime.expose.is_empty());
    assert_eq!(runtime.host, "127.0.0.1");
}

#[test]
fn lan_binds_all_tailscale_keeps_loopback_both_combine() {
    let lan = TodoConfig {
        expose: Some(ExposeValue::Channels(vec![ExposeChannel::Lan])),
        ..Default::default()
    };
    assert_eq!(resolve_runtime_config(&env(&[]), &lan).host, "0.0.0.0");

    let ts = TodoConfig {
        expose: Some(ExposeValue::Channels(vec![ExposeChannel::TailscaleServe])),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[]), &ts);
    assert_eq!(runtime.host, "127.0.0.1");
    assert_eq!(runtime.expose, vec![ExposeChannel::TailscaleServe]);

    let both = TodoConfig {
        expose: Some(ExposeValue::Channels(vec![
            ExposeChannel::Lan,
            ExposeChannel::TailscaleServe,
        ])),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[]), &both);
    assert_eq!(runtime.host, "0.0.0.0");
    assert_eq!(runtime.expose.len(), 2);
}

#[test]
fn env_expose_is_comma_separated_and_wins_entirely() {
    let config_on = TodoConfig {
        expose: Some(ExposeValue::Channels(vec![ExposeChannel::Lan])),
        ..Default::default()
    };
    // env "off"(유효 채널 없음)로 config 를 통째로 덮어 강제 차단
    let runtime = resolve_runtime_config(&env(&[("ROCKY_TODO_EXPOSE", "off")]), &config_on);
    assert!(runtime.expose.is_empty());
    assert_eq!(runtime.host, "127.0.0.1");

    let runtime = resolve_runtime_config(
        &env(&[("ROCKY_TODO_EXPOSE", "lan, tailscale-serve")]),
        &TodoConfig::default(),
    );
    assert_eq!(runtime.expose.len(), 2);
    assert_eq!(runtime.host, "0.0.0.0");
}

#[test]
fn statusline_template_default() {
    let runtime = resolve_runtime_config(&env(&[]), &TodoConfig::default());
    assert_eq!(runtime.statusline_template, DEFAULT_STATUSLINE_TEMPLATE);
}

#[test]
fn statusline_template_config_over_default() {
    let todo = TodoConfig {
        statusline_template: Some("[{doing}]".into()),
        ..Default::default()
    };
    assert_eq!(
        resolve_runtime_config(&env(&[]), &todo).statusline_template,
        "[{doing}]"
    );
}

#[test]
fn statusline_template_env_over_config() {
    let todo = TodoConfig {
        statusline_template: Some("[{doing}]".into()),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[("ROCKY_TODO_STATUSLINE", "[{stale}]")]), &todo);
    assert_eq!(runtime.statusline_template, "[{stale}]");
}

#[test]
fn blank_statusline_template_falls_back_to_default() {
    let todo = TodoConfig {
        statusline_template: Some("   ".into()),
        ..Default::default()
    };
    let runtime = resolve_runtime_config(&env(&[("ROCKY_TODO_STATUSLINE", "  ")]), &todo);
    assert_eq!(runtime.statusline_template, DEFAULT_STATUSLINE_TEMPLATE);
}

// ── load_todo_config (rocky-config.test.ts) ──

fn write_config(content: &str) -> (tempfile::TempDir, std::path::PathBuf) {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("rocky.json");
    std::fs::write(&path, content).unwrap();
    (dir, path)
}

#[test]
fn missing_file_is_fail_open() {
    let config = load_todo_config(std::path::Path::new("/no/such/rocky.json"));
    assert!(config.port.is_none());
}

#[test]
fn unparsable_json_is_fail_open() {
    let (_dir, path) = write_config("not json");
    let config = load_todo_config(&path);
    assert!(config.port.is_none());
}

#[test]
fn non_object_top_level_is_fail_open() {
    let (_dir, path) = write_config("[1,2,3]");
    assert!(load_todo_config(&path).port.is_none());
}

#[test]
fn missing_todo_block_is_fail_open() {
    let (_dir, path) = write_config(r#"{"openapi":{}}"#);
    assert!(load_todo_config(&path).port.is_none());
}

#[test]
fn non_object_todo_is_fail_open() {
    let (_dir, path) = write_config(r#"{"todo":"yes"}"#);
    assert!(load_todo_config(&path).port.is_none());
}

#[test]
fn reads_port_dir_expose_watch() {
    let (_dir, path) =
        write_config(r#"{"todo":{"port":9000,"dir":"/data","expose":["lan"],"watch":false}}"#);
    let config = load_todo_config(&path);
    assert_eq!(config.port, Some(9000));
    assert_eq!(config.dir.as_deref(), Some("/data"));
    assert!(
        matches!(config.expose, Some(ExposeValue::Channels(ref c)) if c == &vec![ExposeChannel::Lan])
    );
    assert_eq!(config.watch, Some(false));
}

#[test]
fn expose_string_and_off_forms() {
    let (_dir, path) = write_config(r#"{"todo":{"expose":"lan"}}"#);
    assert!(matches!(
        load_todo_config(&path).expose,
        Some(ExposeValue::Channels(ref c)) if c == &vec![ExposeChannel::Lan]
    ));
    let (_dir2, path2) = write_config(r#"{"todo":{"expose":"off"}}"#);
    assert!(matches!(
        load_todo_config(&path2).expose,
        Some(ExposeValue::Off)
    ));
    let (_dir3, path3) = write_config(r#"{"todo":{"expose":null}}"#);
    assert!(matches!(
        load_todo_config(&path3).expose,
        Some(ExposeValue::Off)
    ));
}

#[test]
fn wrong_typed_fields_are_dropped() {
    let (_dir, path) = write_config(r#"{"todo":{"port":"9000","dir":123}}"#);
    let config = load_todo_config(&path);
    assert!(config.port.is_none());
    assert!(config.dir.is_none());
}

#[test]
fn reads_statusline_template() {
    let (_dir, path) = write_config(r#"{"todo":{"statusline":{"template":"[{doing}]"}}}"#);
    assert_eq!(
        load_todo_config(&path).statusline_template.as_deref(),
        Some("[{doing}]")
    );
}

#[test]
fn malformed_statusline_is_ignored() {
    let (_dir, path) = write_config(r#"{"todo":{"statusline":"template"}}"#);
    assert!(load_todo_config(&path).statusline_template.is_none());
    let (_dir2, path2) = write_config(r#"{"todo":{"statusline":{"template":123}}}"#);
    assert!(load_todo_config(&path2).statusline_template.is_none());
}
